clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala

```diff
@@ -1,63 +1,254 @@
+    */
+  private val beforeCommandMessages: Messages[CommonOptions] =
+    CommonOptions.messages.copy(
+      appName = "Clio Client",
+      appVersion = ClioClientConfig.Version.value,
+      progName = ClioClient.progName,
+      optionsDesc = "[options] [command] [command-options]"
+    )
+
+  /** Names of all valid sub-commands. */
+  private val commands: Seq[String] =
+    ClioCommand.messages.messagesMap.keys.toSeq
+
+  /** Top-level help message to display on --help. */
+  private val helpMessage: String =
+    s"""
+       |${beforeCommandMessages.helpMessage}
+       |Available commands:
+       |${commands.map(commandHelp).mkString("\n")}
+     """.stripMargin
+
+  /**
+    * Top-level usage message to display on --usage,
+    * or when users give no sub-command.
+    */
+  private val usageMessage: String =
+    s"""
+       |${beforeCommandMessages.usageMessage}
+       |Available commands:
+       |  ${commands.mkString(", ")}
+     """.stripMargin
+
+  /** Help message for a specific command. */
+  private def commandHelp(command: String): String = {
+    ClioCommand.messages
+      .messagesMap(command)
+      .helpMessage(s"${beforeCommandMessages.progName} [options]", command)
   }
 
-  implicit val system: ActorSystem = ActorSystem(progName)
-  import system.dispatcher
-  sys.addShutdownHook({ val _ = system.terminate() })
+  /** Usage message for a specific command. */
+  private def commandUsage(command: String): String = {
+    ClioCommand.messages
+      .messagesMap(command)
+      .usageMessage(
+        s"${beforeCommandMessages.progName} [options]",
+        s"$command [command-options]"
+      )
+  }
 
-  val webClient: ClioWebClient = new ClioWebClient(
-    ClioClientConfig.ClioServer.clioServerHostName,
-    ClioClientConfig.ClioServer.clioServerPort,
-    ClioClientConfig.ClioServer.clioServerUseHttps,
-    ClioClientConfig.responseTimeout
-  )
+  /**
+    * The client's entry point.
+    *
+    * Note: this can't be named `main` because if it is, the compiler will
+    * try to be helpful and generate a forwarder method also named `main`
+    * in the companion object that will do nothing but call this method
+    * (allowing this class to serve as a top-level entry-point), conflicting
+    * with the main method already present in the companion object.
+    */
+  def instanceMain(
+    args: Array[String]
+  ): Either[EarlyReturn, Future[HttpResponse]] = {
+    val maybeParse =
+      ClioCommand.parser.withHelp
+        .detailedParse(args)(CommonOptions.parser.withHelp)
+        .left
+        .map(ParsingError.apply)
 
-  override def run(command: CommandType, remainingArgs: RemainingArgs): Unit = {
+    maybeParse.flatMap {
+      case (commonParse, commonArgs, maybeCommandParse) => {
+        for {
+          _ <- messageIfAsked(helpMessage, commonParse.help)
+          _ <- messageIfAsked(usageMessage, commonParse.usage)
+          commonOpts <- commonParse.baseOrError.left.map(ParsingError.apply)
```

Comment:
This projection to the left happens often enough that I'm wondering if its worth it to pull it out into its own imported implicit class. Ex:
```scala
  implicit class ProjectParsingErrors(either: Either[String, CommonOptions]) {
    def parsingErrorProjection: Either[ParsingError, CommonOptions] = 
      either.left.map(ParsingError.apply)
  }
```
Not necessary, but it would make changing the way parsing error messages are handled easier.

Reply:
I tried adding an implicit conversion from `Either[String, X]` to `Either[EarlyReturn, X]`, but type inference wasn't smart enough to figure out all the places where it should be applied. I'll make an explicit method to wrap around the spots that need it.